### PR TITLE
Enable custom cni provider support for classy clusters

### LINKED ISSUE #4292: Clusterbootstrap  logic does not allow for custom CNI Provider
**Bug description**
clusterboostrap webhooks and validators do not allow the spec.CNI of  a clusterbootstrap to be nil. 
This is a  problem because when using a custom CNI provider, the spec.CNI of clusterbootstrap is set to nil. 

**Affected product area (please put an X in all that apply)**

- ( ) APIs
- ( x) Addons
- ( ) CLI
- ( ) Docs
- ( ) IAM
- ( ) Installation
- ( ) Plugin
- ( ) Security
- ( ) Test and Release
- ( ) User Experience
- ( ) Developer Experience

**Expected behavior**
The clusterbootstrap webhooks and validators should allow clusterbootstrap.spec.cni = nil in the case where a custom CNI provider is being used
**Steps to reproduce the bug**

**Version** (include the SHA if the version is not obvious)

**Environment where the bug was observed (cloud, OS, etc)**

**Relevant Debug Output (Logs, manifests, etc)**

### PR DESCRIPTION
**Pull request by adduarte**
Adds logic to handle tkg.tanzu.vmware.com/use-custom-cni-provider annotation.

### What this PR does / why we need it
The case when clusterboostrap.spec.cni = nil needs to be supported because it is how tanzu supports
the use of a custom CNI provider. 

When the value of clusterboostrap.spec.cni == nil, the CNI infrastructure is not configured. 
The CNI is expected to be configured by the administrator. 

This PR adds logic to handle a new annotation in a clusterbootstrap CR. 

if the clusterboostrap CR is annotated with: 

tkg.tanzu.vmware.com/use-custom-cni-provider: ""

CNI for the cluster will not be configured and packages will not be installed. 

clusterbootstrap.spec.cni must be set to nil. 

annotating a clusterbootstrap and setting spec.cni to something else than nil will result in a validation error.

### Which issue(s) this PR fixes

Fixes #4292

### Describe testing done for PR

Unit tests provided for new logic. 
deployment of addons-manager image with fix deployed to existing aws cluster. 
Tested on vsphere cluster and verified cni unmanaged functionality. 

### Release note

```release-note
annotating a clusterbootstrap CR with tkg.tanzu.vmware.com/unmanaged-cni: "", will allow for the clusterbootstrap.spec.cni = nil
```

### CODE REVIEW

File: addons/pkg/constants/constants.go [hunk 1]
```diff
@@ -282,6 +282,9 @@ const (
 	// logic
 	AddCBMissingFieldsAnnotationKey = "tkg.tanzu.vmware.com/add-missing-fields-from-tkr"
 
+	// UseCustomCNIProvider is the label for clusters that use a custom CNI provider
+	UseCustomCNIProvider = "tkg.tanzu.vmware.com/use-custom-cni-provider"
```
Comment (shyaamsn) [inline addons/pkg/constants/constants.go:286]: 1. In validation webhook check if spec.cni is nil if this annotation is given
2. In defaulting webhook skip adding spec.cni from CBT if this annotation is given
3. Fix validating webhooks where spec.cni is mandatory.

Comment (adduarte) [reply addons/pkg/constants/constants.go:286]: checklist done

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go [hunk 2]
```diff
@@ -62,6 +63,10 @@ func NewHelper(ctx context.Context, k8sClient client.Client, aggregateAPIResourc
 // a fully qualified name(<packageShortName>.<domain>.<version>). The fully qualified name is fetched from TKR.
 func (h *Helper) CompleteCBPackageRefNamesFromTKR(tkr *runtanzuv1alpha3.TanzuKubernetesRelease, clusterBootstrap *runtanzuv1alpha3.ClusterBootstrap) error {
 	var suffix = "*"
+
+	if clusterBootstrap.Spec == nil {
+		return errors.New("missing spec definition")
+	}
 	clusterBootstrapPackages := []*runtanzuv1alpha3.ClusterBootstrapPackage{
```
Comment (adduarte) [inline addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go:70]: hmm. I think Codecov is missing something. This lines do get cover by this test: 
https://github.com/vmware-tanzu/tanzu-framework/pull/4264/files#diff-2d52a425f83e5f6840a1829491d52865109c05aa05a320f36b1eb397c784f53cR487

"Clusterbootstrap is NOT annotated and spec is nil"

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go [hunk 3]
```diff
@@ -62,6 +63,10 @@ func NewHelper(ctx context.Context, k8sClient client.Client, aggregateAPIResourc
 // a fully qualified name(<packageShortName>.<domain>.<version>). The fully qualified name is fetched from TKR.
 func (h *Helper) CompleteCBPackageRefNamesFromTKR(tkr *runtanzuv1alpha3.TanzuKubernetesRelease, clusterBootstrap *runtanzuv1alpha3.ClusterBootstrap) error {
 	var suffix = "*"
+
+	if clusterBootstrap.Spec == nil {
```
Comment (adduarte) [inline addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go:67]: This is not strictly related to the issue this PR solves, but it did run into null pointer error when writing unit tests, so best to add this fix here.

**Comment by alfredthenarwhal** [comment]
> @adduarte:  /test install-vc7
Commit: 74661de

File: addons/webhooks/clusterbootstrap_webhook_test.go [hunk 4]
```diff
@@ -404,3 +435,244 @@ func assertFindKubeVipInClusterBootstrap(clusterBootstrap *runv1alpha3.ClusterBo
 	Expect(kubevipPackage.RefName).To(Equal(fmt.Sprintf("%s.%s", fakeKubevipcloudproviderCarvelPackageRefName, fakeCarvelPackageVersion)))
 	Expect(kubevipPackage.ValuesFrom.Inline["foo"]).To(Equal("bar"))
 }
+
+var _ = Describe("Custom CNI Provider:", func() {
```
Comment (shyaamsn) [inline addons/webhooks/clusterbootstrap_webhook_test.go:439]: Nice job on covering all the unit tests

File: addons/pkg/constants/constants.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (knabben) [reply addons/pkg/constants/constants.go:286]: maybe something like `skip-cni-installation`

[the remark this answers is not in the document]
Comment (shyaamsn) [reply addons/pkg/constants/constants.go:286]: We do installation , upgrade as well. 

My vote is for `tkg.tanzu.vmware.com/unmanaged-cni`

File: addons/packages/addons-manager/1.4.0/bundle/config/upstream/addons-manager.yaml [hunk 5]
```diff
@@ -11,6 +11,13 @@ kind: ClusterRole
 metadata:
   name: tanzu-addons-manager-clusterrole
 rules:
+- apiGroups:
```
[the remark this answers is not in the document]
Comment (adduarte) [reply addons/packages/addons-manager/1.4.0/bundle/config/upstream/addons-manager.yaml:14]: ack. reverting change

File: addons/controllers/clusterbootstrap_controller.go [hunk 6]
```diff
@@ -500,19 +500,23 @@ func (r *ClusterBootstrapReconciler) mergeClusterBootstrapPackagesWithTemplate(
...
-		// We don't allow change to the CNI selection once it starts running, however we allow version bump
-		// TODO: check correctness of the following statement, as we still allow version bump
-		// ClusterBootstrap webhook will make sure the package RefName always match the original CNI
-		updatedCNI, cniNamePrefix, err := util.GetBootstrapPackageNameFromTKR(r.context, r.Client, updatedClusterBootstrap.Spec.CNI.RefName, cluster)
-		if err != nil {
-			errorMsg := fmt.Sprintf("unable to find any CNI bootstrap package prefixed with '%s' for ClusterBootstrap %s/%s in TKR", cniNamePrefix, cluster.Namespace, cluster.Name)
-			return nil, errors.Wrap(err, errorMsg)
+
+	_, usingCustomCNIProvider := updatedClusterBootstrap.Annotations[constants.UseCustomCNIProvider]
+	if !(usingCustomCNIProvider && updatedClusterBootstrap.Spec.CNI == nil) {
```
[the remark this answers is not in the document]
Comment (adduarte) [reply addons/controllers/clusterbootstrap_controller.go:505]: The requirements to indicate that the user wants to use a custom cni is to annotate the clusterbootstrap *and* that Spec.CNI == nil. 

So the opposite (using a built in provider like antrea or calico) is the oposite of that: 
  
!(usingCustomCNIProvider && updatedClusterBootstrap.Spec.CNI == nil) = usingBuiltInProvider (new variable I will introduce to make clear). 

The inside check (updatedClusterBootstrap.Spec.CNI == nil ) is done because there is a chance that usingCustomCNIProvider = false && Spec.CNI== nil, in which case we will enter the loop. 

Will reformat code to make this clearer.

File: addons/pkg/constants/constants.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (adduarte) [reply addons/pkg/constants/constants.go:286]: changing to tkg.tanzu.vmware.com/unmanaged-cni

File: addons/controllers/clusterbootstrap_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (adduarte) [reply addons/controllers/clusterbootstrap_controller.go:505]: done

[the remark this answers is not in the document]
Comment (adduarte) [reply addons/controllers/clusterbootstrap_controller.go:505]: done

File: addons/pkg/constants/constants.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (adduarte) [reply addons/pkg/constants/constants.go]: done

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 7]
```diff
@@ -168,8 +178,16 @@ func (wh *ClusterBootstrap) ValidateCreate(ctx context.Context, obj runtime.Obje
 	var allErrs field.ErrorList
 	// Iterating one by one because we need field info for getFieldPath, and some core packages can be nil
 
-	if err := wh.validateClusterBootstrapPackage(ctx, clusterBootstrap.Spec.CNI, clusterBootstrap.Namespace, getFieldPath("cni")); err != nil {
-		allErrs = append(allErrs, err)
+	// If  spec.cni == nil and clusterbootstrap is  annotated,  it is using unmanaged CNI, and we do not need to validate CNI
+	_, usingUnmanagedCNI := clusterBootstrap.Annotations[constants.UseUnmanagedCNI]
```
Comment (shyaamsn) [inline addons/webhooks/clusterbootstrap_webhook.go:182]: ValidateUpdate is missing the same checks ?

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 8]
```diff
@@ -136,6 +136,12 @@ func (wh *ClusterBootstrap) Default(ctx context.Context, obj runtime.Object) err
 	// Get the helper ready for the defaulting logic
 	helper := clusterbootstrapclone.Helper{Logger: clusterbootstraplog}
 
+	_, usingCustomCNIProvider := clusterBootstrap.Annotations[constants.UseCustomCNIProvider]
+
+	if usingCustomCNIProvider && clusterBootstrap.Spec.CNI != nil {
```
[the remark this answers is not in the document]
Comment (adduarte) [reply addons/webhooks/clusterbootstrap_webhook.go]: Added to both default and validator because if the user submits a clusterbootstrap with the AddCBMissingFieldsAnnotationKey and UnmanagedCNI annotations and the spec.cni != nil, it is difficult to interpret since UnmangeCNI means spec.cni = nil, but the user did provide one.
We could make the decision for user and simply set spec.cni = nil whenever we see unmanagecni regardless of the user provided value for spec.cni

File: addons/webhooks/clusterbootstrap_webhook.go (hunk 7 above)
Comment (adduarte) [reply addons/webhooks/clusterbootstrap_webhook.go]: added checks to validateUpdate
done

File: addons/webhooks/clusterbootstrap_webhook.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (adduarte) [reply addons/webhooks/clusterbootstrap_webhook.go]: we will move ahead with setting spec.nil == nil if clusterbootstrap is annotated. (ignoring user provided spec.cni because this is is a default webhook, so we will take the annotation presence as a request of user to fill in spec.cni for them).

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go [hunk 9]
```diff
@@ -368,6 +373,10 @@ func (h *Helper) HandleExistingClusterBootstrap(clusterBootstrap *runtanzuv1alph
 			}
 		}
 
+		if _, ok := clusterBootstrap.Annotations[constants.UnmanagedCNI]; ok {
+			clusterBootstrapTemplate.Spec.CNI = clusterBootstrap.Spec.CNI
```
Comment (knabben) [inline addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go:377]: why to we need to set the CBT.spec.CNI when the unmanagedCNI is used?

**Review by knabben — APPROVED** [review]

File: addons/webhooks/clusterbootstrap_webhook.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (adduarte) [reply addons/webhooks/clusterbootstrap_webhook.go]: done

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go (hunk 9 above)
Comment (adduarte) [reply addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go:377]: in order to leave the CNI part untouched (unmanaged) it's easier to set the templateCB.Sepc.CNI  to the same value of the currentCB.Spec.CNI before making the call to AddMissingSpecFieldsFromTemplate.  
That way currentCB.Spec.CNI stays that same.  Otherwise we would have to heavily modify AddMissingSpecFieldsFromTemplate to handle the case where clusterBootstrap.Annotations[constants.UnmanagedCNI]

We could set templateCB.Spec.CNI = nil but since technically "unmanaged" means we don't change it, we leave the setting to nil to the default webhook, and take the full meaning of unmanaged, which means we don't change it at all.

**Review by shyaamsn — APPROVED** [review]
LGTM
